velox/functions/prestosql/ArraySum.cpp

```diff
@@ -0,0 +1,216 @@
+    const TypePtr& outputType,
+    exec::EvalCtx* context,
+    VectorPtr* result) const {}
+
+template <>
+void ArraySumFunction<StringView, double>::apply(
+    const SelectivityVector& rows,
+    std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+    const TypePtr& outputType,
+    exec::EvalCtx* context,
+    VectorPtr* result) const {}
+
+template <>
+void ArraySumFunction<Date, int64_t>::apply(
+    const SelectivityVector& rows,
+    std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+    const TypePtr& outputType,
+    exec::EvalCtx* context,
+    VectorPtr* result) const {}
+
+template <>
+void ArraySumFunction<Date, double>::apply(
+    const SelectivityVector& rows,
+    std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+    const TypePtr& outputType,
+    exec::EvalCtx* context,
+    VectorPtr* result) const {}
+
+// Validate number of parameters and types.
+void validateType(const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_USER_CHECK_EQ(
+      inputArgs.size(), 1, "array_sum requires exactly one parameter");
+
+  auto arrayType = inputArgs.front().type;
+  VELOX_USER_CHECK_EQ(
+      arrayType->kind(),
+      TypeKind::ARRAY,
+      "array_sum requires argument of type ARRAY");
+
+  auto valueTypeKind = inputArgs.front().type->childAt(0)->kind();
+  bool isCoercibleToDouble = valueTypeKind == TypeKind::TINYINT ||
+      valueTypeKind == TypeKind::SMALLINT ||
+      valueTypeKind == TypeKind::INTEGER || valueTypeKind == TypeKind::BIGINT ||
+      valueTypeKind == TypeKind::REAL || valueTypeKind == TypeKind::DOUBLE;
+  VELOX_USER_CHECK_EQ(isCoercibleToDouble, true, "Invalid value type");
+}
+
+// Create function template based on type.
+template <TypeKind kind>
+std::shared_ptr<exec::VectorFunction> createTyped(
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_CHECK_EQ(inputArgs.size(), 1);
+
+  using IT = typename TypeTraits<kind>::NativeType;
+  if (kind == TypeKind::TINYINT || kind == TypeKind::SMALLINT ||
+      kind == TypeKind::INTEGER || kind == TypeKind::BIGINT) {
+    return std::make_shared<ArraySumFunction<IT, int64_t>>();
+  }
+  if (kind == TypeKind::REAL || kind == TypeKind::DOUBLE) {
+    return std::make_shared<ArraySumFunction<IT, double>>();
+  }
+  VELOX_FAIL("Unsupported type")
+}
+
+// Create function.
+std::shared_ptr<exec::VectorFunction> create(
+    const std::string& /* name */,
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  validateType(inputArgs);
+  auto elementType = inputArgs.front().type->childAt(0);
+
+  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      createTyped, elementType->kind(), inputArgs);
+}
+
+// Define function signature.
+// array(T1) -> T2 where T1 must be coercible to bigint or double, and
+// T2 is bigint or double
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
```

Comment:
The `signatures()` approach from `ArrayMinMax.cpp` can be followed here as well, using an `unordered_map` instead of the `vector`.